### PR TITLE
Feature: expand docs for `blackpill-f4`

### LINKED ISSUE #1899: Not working with default crossfile
Hey, I am using default crossfile for configuration like this " meson setup build --cross-file=cross-file/blackpill-f401cc.ini " and I am able to compile and generate a .bin file as well but it does not list itself as a USB device in device manager. Is it possible that this is happening because of not having correct dependencies? I am using a blackpill board(401cc) so I have installed dependencies for 3rd party hardware using " sudo apt install libusb-1.0-0-dev libftdi1-dev libhidapi-dev ". I am using WSL(Ubuntu22) and stlink utility for flashing bmp firmware using clone stlink.
Also after configuration in few places I can see, it says few fiiles are not found like ' bmputil found: NO' and few warnings "libopencm3 : YES 2 warnings" are these okay? I had done this on an Ubuntu machine it was not showing anything like this and my BMP was also working fine,  probably because it had some other dependencies already installed that I am not aware of. I could not find any better place to ask this question please guide me on this.

picture of what I see in device manager when BMP is connected :
Screenshot 2024-08-17 062030
adding some screenshots of what I see after configuration :
image
image

### PR DESCRIPTION
**Pull request by ALTracer**
## Detailed description

* This is a UX feature, no code is changed.
* The existing problem is unobvious workflow of blank board flash upgrade for blackpill-f4.
* This PR helps solve it by expanding documentation in platform README with Meson-specific and BMD-bootloader-specific paragraphs.

This partially reverts bb6a7ce to keep Makefile docs portion, because both buildsystem technically still work, and there is no counterpart to lenvm's SHIELD=1 makeflag in meson flags.
I feel like BMDbootloader-specific portion of docs is board-agnostic, but there is no good existing Markdown file to move this into and cross-link.
Awaiting stylistic review of most text, as English is not my native language (I know the vocabulary enough, but not the syntax details).
 
## Your checklist for this pull request

* [x] I've read the [Code of Conduct](https://github.com/blackmagic-debug/blackmagic/blob/main/CODE_OF_CONDUCT.md)
* [x] I've read the [guidelines for contributing](https://github.com/blackmagic-debug/blackmagic/blob/main/CONTRIBUTING.md) to this repository
* [x] It builds for hardware native (see [Building the firmware](https://github.com/blackmagic-debug/blackmagic?tab=readme-ov-file#building-black-magic-debug-firmware))
* [x] It builds as BMDA (see [Building the BMDA](https://github.com/blackmagic-debug/blackmagic?tab=readme-ov-file#building-black-magic-debug-app))
* [x] I've tested it to the best of my ability
* [x] My commit messages provide a useful short description of what the commits do

## Closing issues

Fixes #1899.

### CODE REVIEW

**Review by dragonmux — CHANGES_REQUESTED** [review]
This generally looks very good. There are some items to change regarding the Git instructions and build directory naming, plus a couple of wording suggestions in the final section, but nothing show-stopping or major. With them taken care of, we think this'll be good to merge.

File: src/platforms/common/blackpill-f4/README.md [hunk 1]
```diff
@@ -81,3 +82,57 @@ mon freq 900k
...
+```
+
+1. Create a build configuration for specific platform (WeActStudio MiniSTM32F401CC) with specific options (enable BMD bootloader flags and offsets -- this is the default)
+
+`$ meson setup build-f401cc --cross-file=cross-file/blackpill-f401cc.ini -Dbmd_bootloader=true`
+
+2. Compile the firmware and bootloader
+
+```sh
+$ ninja -C build-f401cc/ blackmagic_blackpill_f401cc_firmware.bin
```
[the remark this answers is not in the document]
Comment (ALTracer) [reply src/platforms/common/blackpill-f4/README.md:105]: Ah, but notice I describe two build variants in separate dirs. If the user is to try the next variant in the same dir, then meson prints a notice and changes nothing. And I can't recommend `meson configure` either, it has to be `meson setup` first time.
What about `meson compile -C build`?

[the remark this answers is not in the document]
Comment (dragonmux) [reply src/platforms/common/blackpill-f4/README.md:105]: A new user isn't going to be trying multiple variants. There are already instructions to be written more generally about Meson usage that will cover switching a build directory between probes, so please presume here a new user coming to this for the first time and probably wanting to contribute. We need to account for the fact they must not add the build directory contents to version control and if they do what you're suggesting here, and then `git add -a`, then that's just what they'll do.

File: src/platforms/common/blackpill-f4/README.md [hunk 2]
```diff
@@ -81,3 +82,57 @@ mon freq 900k
...
+* ST MaskROM is the read-only System Memory bootloader which starts up per BOOT0-triggered SYSCFG mem-remap. It talks USART AN2606 so you can use stm32flash etc. When USBOTG_FS is plugged, it spins up HSE 25 MHz and measures its frequency relative to HSI (RC 8 MHz with worse tolerance) and sometimes can misdetect it as 24 or 26 MHz then configure a wrong input PLL divider, resulting in broken USB DFU. New 8 MHz HSE boards should not be affected by this, but also they are not yet supported (needs a different pll config).
+* BMD bootloader is the port of BMP bootloader with dfu_f1.c internal flash driver swapped to dfu_f4.c (and different serialno algorithm etc.) which a) has a fixed per-board PLL config, no autodetection; b) understands buttons, drives LED, does not touch other GPIOs, talks USB DfuSe, ~~has MS BOS WINUSB descriptors~~, uses same libopencm3 code so you can verify hardware config via a smaller binary; c) erases and writes to internal Flash ~2.4x faster than MaskROM; d) all of that in first 8-9 KiB of first page of 16 KiB (of F2/F4/F7 flash), just like on `native`/`stlink`/`swlink` etc.
```
[the remark this answers is not in the document]
Comment (ALTracer) [reply src/platforms/common/blackpill-f4/README.md:138]: Accepted. Does it? Current `dfu_init()` only calls `usbd_init()`. It also has to call that https://github.com/blackmagic-debug/blackmagic/blob/488bd37ca00d6cb2578e96d88af9a9df9edbb636/src/platforms/common/usb.c#L59-L60 and have descriptor structs expanded.
The paragraph is too wordy and unstructured, I'll review it in ReText, even if it doesn't have a CommonMark mode.

[the remark this answers is not in the document]
Comment (dragonmux) [reply src/platforms/common/blackpill-f4/README.md:138]: Indeed, the bootloader does not yet support this, which is why we're not asking you to remove the strike-through - the wording nitpick is that the BOS (Binary Object Storage) is not MS specific, and the descriptors are not WINUSB descriptors but rather MS OS Descriptors that define bindings for WINUSB to interfaces. It's perfectly possible to bind other drivers to interfaces instead of WINUSB if we wanted, among other things those descriptors have the power to do.

[the remark this answers is not in the document]
Comment (ALTracer) [reply src/platforms/common/blackpill-f4/README.md]: Ironically, this is now a ST side benefit, they get to ship signed drivers for their "ST BOOTLOADER", meanwhile BMD bootloader requires manual drivers association on anything Windows.

File: src/platforms/common/blackpill-f4/README.md [hunk 3]
```diff
@@ -81,3 +82,74 @@ mon freq 900k
...
+## Instructions for Meson buildsystem
+
+### Variant with meson and BMD bootloader enabled (default)
+
+0. Clone the repo and libopencm3 submodule, install toolchains, meson, etc.
+
+```sh
+$ git clone https://github.com/blackmagic-debug/blackmagic.git
+$ cd blackmagic
+$ git submodule update --init
```
Comment (dragonmux) [inline src/platforms/common/blackpill-f4/README.md:108]: The style of shell blocks in the documentation, in part to make copying commands into the console easier, is to not include a `$ ` prefix, so please drop that prefix on all shell blocks here and below. Even when the block is for just one command, please use triple-backticks blocks marked for shell on them, this keeps the presentation style consistent.

**Review by dragonmux — COMMENTED** [review]
Once the one remaining stylistic issue has been corrected, we're happy to merge this - LGTM save for that one thing.

**Review by dragonmux — APPROVED** [review]
LGTM, merging. Thank you for the contribution!
